server/src/main/java/org/apache/cassandra/sidecar/modules/RestoreJobModule.java

```diff
@@ -0,0 +1,195 @@
+import com.google.inject.Singleton;
+import com.google.inject.TypeLiteral;
+import com.google.inject.multibindings.MapBinder;
+import com.google.inject.multibindings.ProvidesIntoMap;
+import org.apache.cassandra.sidecar.cluster.InstancesMetadata;
+import org.apache.cassandra.sidecar.cluster.locator.CachedLocalTokenRanges;
+import org.apache.cassandra.sidecar.cluster.locator.LocalTokenRangesProvider;
+import org.apache.cassandra.sidecar.common.server.dns.DnsResolver;
+import org.apache.cassandra.sidecar.config.SidecarConfiguration;
+import org.apache.cassandra.sidecar.db.schema.RestoreJobsSchema;
+import org.apache.cassandra.sidecar.db.schema.RestoreRangesSchema;
+import org.apache.cassandra.sidecar.db.schema.RestoreSlicesSchema;
+import org.apache.cassandra.sidecar.db.schema.TableSchema;
+import org.apache.cassandra.sidecar.handlers.DiskSpaceProtectionHandler;
+import org.apache.cassandra.sidecar.handlers.RouteBuilder;
+import org.apache.cassandra.sidecar.handlers.restore.AbortRestoreJobHandler;
+import org.apache.cassandra.sidecar.handlers.restore.CreateRestoreJobHandler;
+import org.apache.cassandra.sidecar.handlers.restore.CreateRestoreSliceHandler;
+import org.apache.cassandra.sidecar.handlers.restore.RestoreJobProgressHandler;
+import org.apache.cassandra.sidecar.handlers.restore.RestoreJobSummaryHandler;
+import org.apache.cassandra.sidecar.handlers.restore.RestoreRequestValidationHandler;
+import org.apache.cassandra.sidecar.handlers.restore.UpdateRestoreJobHandler;
+import org.apache.cassandra.sidecar.handlers.validations.ValidateTableExistenceHandler;
+import org.apache.cassandra.sidecar.modules.multibindings.ClassKey;
+import org.apache.cassandra.sidecar.modules.multibindings.KeyClassMapKey;
+import org.apache.cassandra.sidecar.modules.multibindings.PeriodicTaskMapKeys;
+import org.apache.cassandra.sidecar.modules.multibindings.TableSchemaMapKeys;
+import org.apache.cassandra.sidecar.modules.multibindings.VertxRouteMapKeys;
+import org.apache.cassandra.sidecar.restore.RestoreJobDiscoverer;
+import org.apache.cassandra.sidecar.restore.RestoreProcessor;
+import org.apache.cassandra.sidecar.restore.RingTopologyRefresher;
+import org.apache.cassandra.sidecar.routes.VertxRoute;
+import org.apache.cassandra.sidecar.tasks.PeriodicTask;
+
+/**
+ * Provides the capability of restoring data (SSTables) from S3 into Cassandra
+ */
+public class RestoreJobModule extends AbstractModule
+{
+    @Override
+    protected void configure()
+    {
+        // Leverage TypeLiteral to preserve the lower bound type Key at the runtime
+        TypeLiteral<Class<? extends ClassKey>> keyType = new TypeLiteral<>() {};
+        TypeLiteral<PeriodicTask> valueType = new TypeLiteral<>() {};
+        MapBinder<Class<? extends ClassKey>, PeriodicTask> periodicTaskMapBinder = MapBinder.newMapBinder(binder(), keyType, valueType);
+        periodicTaskMapBinder.addBinding(PeriodicTaskMapKeys.RestoreJobDiscovererKey.class).to(RestoreJobDiscoverer.class);
```

Comment:
Curious, why do we inject for these periodic tasks and bind them this way. Where as for `HealthCheckPeriodicTask` we provide with `ProvidesIntoMap`

Reply:
They are equivalent. I wanted to demo that there are different ways to add bindings. Plus, binding to a class (that is annotated with @Inject) has simpler code.